#FINRA PRE-SCREENING TASK 
##By HARSHA DESHPANDE
Date: 01-07-2019

This Notebook contains prelimnary tasks required to qualify for an Inteview for the poition of Business Data Analyst for FINRA -- Financial Industry Regulatory Authority, Inc.

In [2]:
%sql
DROP TABLE TRADES

In [3]:
%sql
CREATE TABLE TRADES
(
TradeDate DATE,
FIRM varchar(255),
SYMBOL varchar(255),
SIDE char(1),
QUANTITY INTEGER,
PRICE INTEGER
);

        

In [4]:
%sql
INSERT INTO TABLE TRADES select t.* from (SELECT to_date('2014-02-03') ,'1ABC','A123','B', 200, 41) t

In [5]:
%sql
INSERT INTO TABLE TRADES select t.* from (select to_date('2014-02-04'),'2BCD','B234','B', 600, 60) t

In [6]:
%sql
INSERT INTO TABLE TRADES select t.* from (select to_date('2014-02-07'), '1ABC', 'C345', 'S', 600, 70) t

In [7]:
%sql
INSERT INTO TABLE TRADES select t.* from (select to_date('2014-02-10'),'3CDE','C345','S', 600, 70) t

In [8]:
%sql
INSERT INTO TABLE TRADES select t.* from (select to_date('2014-02-12'),'4DEF','B234','B', 200, 62) t

In [9]:
%sql
INSERT INTO TABLE TRADES select t.* from (select to_date('2014-02-14'), '3CDE', 'B234', 'B', 300, 61) t

In [10]:
%sql
INSERT INTO TABLE TRADES select t.* from (select to_date('2014-02-21'),'1ABC','A123','B', 300, 40) t

In [11]:
%sql
INSERT INTO TABLE TRADES select t.* from (select to_date('2014-02-24'),'1ABC','A123','S', 300, 30) t

In [12]:
%sql
INSERT INTO TABLE TRADES select t.* from (select to_date('2014-02-25'),'4DEF','C345','B', 2100, 71) t

In [13]:
%sql
INSERT INTO TABLE TRADES select t.* from (select to_date('2014-02-27'),'CDE','B234','S', 1100, 63) t

In [14]:
%sql
SELECT * FROM TRADES

TradeDate FIRM SYMBOL SIDE QUANTITY PRICE 2014-02-25 4DEF C345 B 2100 71 2014-02-12 4DEF B234 B 200 62 2014-02-04 2BCD B234 B 600 60 2014-02-10 3CDE C345 S 600 70 2014-02-24 1ABC A123 S 300 30 2014-02-07 1ABC C345 S 600 70 2014-02-21 1ABC A123 B 300 40 2014-02-03 1ABC A123 B 200 41 2014-02-14 3CDE B234 B 300 61 2014-02-27 CDE B234 S 1100 63

In [15]:
%sql
DESCRIBE TRADES

col_name data_type comment TradeDate date null FIRM string null SYMBOL string null SIDE string null QUANTITY int null PRICE int null

In [16]:
%sql
SELECT 
      TradeDate,
      QUANTITY 
FROM TRADES
ORDER BY 1

TradeDate QUANTITY 2014-02-03 200 2014-02-04 600 2014-02-07 600 2014-02-10 600 2014-02-12 200 2014-02-14 300 2014-02-21 300 2014-02-24 300 2014-02-25 2100 2014-02-27 1100

In [17]:
%sql
SELECT 
      TradeDate,
      PRICE 
FROM TRADES
ORDER BY 1 

TradeDate PRICE 2014-02-03 41 2014-02-04 60 2014-02-07 70 2014-02-10 70 2014-02-12 62 2014-02-14 61 2014-02-21 40 2014-02-24 30 2014-02-25 71 2014-02-27 63

In [18]:
%sql
SELECT 
      AVG(PRICE) AS AVG_PRICE 
FROM TRADES 

AVG_PRICE 56.8

In [19]:
%sql
SELECT 
      SUM(PRICE) AS SUM_PRICE
FROM TRADES

SUM_PRICE 568

In [20]:
%sql
SELECT 
      MAX(PRICE) AS MAX_PRICE
FROM TRADES

MAX_PRICE 71

In [21]:
%sql
SELECT 
      MIN(PRICE) AS MIN_PRICE
FROM TRADES

MIN_PRICE 30

##Conclusion:
The Trades table contains the fact table for the trades for the month of February for the Year 2014. This is a Time series data. There is no significant linear seasonality or trend for the measures vs time. The transactions are in between the days 3 and 27 for the month. The Dimensions / Facts here are FIRM, SYMBOL and SIDE. FIRM and SYMBOL are Categorical in nature. Comparatively, SIDE is Discrete / Nominal in nature, with values determining Buy/Sell of the trade. The granularity here is determined by the Measures i.e. QUANTITY and PRICE in this case. These are numerical in nature and calculable. The maximum Price is $71 during the month of February. The minimum Price is $30 for the month of February. The average price is $56.8 and a total amount for all the trades is $568.0.

A.Your business user asks you to tell them the unique symbols in the data table above. 

i. Please write the SQL query you would use to query this table.

ii.	Please show the exact results you expect based on your SQL query.

In [24]:
%sql
SELECT 
      DISTINCT SYMBOL
FROM TRADES

SYMBOL C345 A123 B234

B. Your business user asks you to tell them the unique firm and symbol combinations in the data table above.

i. Please write the SQL query you would use to query this table.

ii. Please show the exact results you expect based on your SQL query.

In [26]:
%sql
SELECT
      FIRM,
      SYMBOL
FROM TRADES 
GROUP BY 1, 2 

FIRM SYMBOL 3CDE C345 2BCD B234 CDE B234 4DEF C345 1ABC A123 3CDE B234 4DEF B234 1ABC C345

C. Your business user asks you to show them a table that includes the number of trades for each firm and symbol combination in the data table above.

a. Please write the SQL query you would use to query this table.

b. Please show the exact results you expect based on your SQL query.

In [28]:
%sql
SELECT 
      FIRM,
      SYMBOL,
      count(QUANTITY) AS NUMBER_TRADES
FROM TRADES
GROUP BY 1, 2
ORDER BY 3 DESC

FIRM SYMBOL NUMBER_TRADES 1ABC A123 3 3CDE C345 1 2BCD B234 1 CDE B234 1 4DEF C345 1 3CDE B234 1 4DEF B234 1 1ABC C345 1

Please write the business requirements and corresponding functional specifications for this business request in the following format. 



BR # 1

Business Requirement: 

To create a report that shows, for each firm, 
the symbols they traded and the total dollar value they traded in that symbol for the period of 02/03/2014 to
02/27/2014. Use Pivot table to summarize the report.

Functional Specification: 

Table Name: Trades

This is a transaction data that represents trades in the symbol noted in the table.

Field Names:

TradeDate is the date on which the trade took place. Datatype here is DATE.

Firm is the Broker/Dealer who made the trade. Datatype here is STRING.

Symbol is the security that was traded. Datatype here is STRING.

Side represents whether the trade was a buy (purchase) or sell (sale) of the security. Datatype here is CHARECTER.

Quantity is the number of shares involved in the transaction. Datatype here is INTEGER.

Price is the price per share. Datatype here is INTEGER.

In [30]:
%sql
SELECT 
      FIRM,
      SYMBOL,
      SUM(QUANTITY * PRICE) AS TOTAL_DOLLAR_AMOUNT 
FROM TRADES 
GROUP BY 1, 2
ORDER BY 3 DESC

FIRM SYMBOL TOTAL_DOLLAR_AMOUNT 4DEF C345 149100 CDE B234 69300 1ABC C345 42000 3CDE C345 42000 2BCD B234 36000 1ABC A123 29200 3CDE B234 18300 4DEF B234 12400

##Thank You!